File: derekbanas/js/08_quick_sort.js


function quickSort(inputs) {
    quickSortBody(inputs, 0, inputs.length - 1);
}


function quickSortBody(inputs, left, right) {
    var start = left;
    var end = right;

    if (left < right) {
        var pivot = inputs[right];
        while(left <= right) {
            while(inputs[left] < pivot) {
                left ++;
            }
            while(inputs[right] > pivot) {
                right --;
            }
            if (left <= right) {
                var tmp = inputs[right];
                inputs[right] = inputs[left];
                inputs[left] = tmp;
                left ++;
                right --;
            }
        }
        quickSortBody(inputs, start, right);
        quickSortBody(inputs, left, end);
    }
}


var inputs = [
    1, 23, 4, 434, 232, 324, 11, 2323, 5
];

quickSort(inputs);

console.log(inputs);

File: derekbanas/python/08_quick_sort.py
from recorder import Recorder
record = Recorder()


def quick_sort(inputs):
    quick_sort_body(inputs, 0, len(inputs) - 1)


def quick_sort_body(inputs, left, right):
    start = left
    end = right
    if left < right:
        pivot = inputs[right]
        while left <= right:
            while inputs[left] < pivot:
                left += 1
            while inputs[right] > pivot:
                right -= 1

            if left <= right:
                tmp = inputs[left]
                inputs[left] = inputs[right]
                inputs[right] = tmp
                left += 1
                right -= 1

        record.add(inputs)
        # print "left: %d to %d, right: %d to %d" % (start, right, left, end)
        quick_sort_body(inputs, start, right)
        quick_sort_body(inputs, left, end)


if __name__ == '__main__':
    inputs = [
        1, 23, 4, 434, 232, 324, 11, 2323, 5
    ]
    quick_sort(inputs)
    record.show()

    # first pass: pivot = 5
    # 1, 5, 4, <break here for next pass> 434, 232, 324, 11, 2323, 23



